src/client/networkclient.cc

```diff
@@ -0,0 +1,371 @@
+
+    if (output()) {
+    *output() << "NetworkClient::connect" << endl
+        << "    client interface: " << client_interface_name_ << endl
+        << "    server host: " << server_name_ << endl
+        << "    server port: " << server_port_ << endl;
+    }
+
+    if (1) {
+    // TODO: Why is bind causing a problem here?
+    // if (client_socket_->bind(QHostAddress(client_interface_name))) {
+      if (output()) {
+        *output() << "NetworkClient: bind successful." << endl;
+      }
+      client_socket_->connectToHost(server_name_, server_port_);
+      setConnectionStatus(ConnectionStatus::Connecting);
+    } else {
+      if (output()) {
+        *output() << "NetworkClient: bind failed." << endl;
+      }
+    }
+  }
+}
+
+void NetworkClient::disconnect() {
+  if (output()) {
+    *output() << "NetworkClient: Disconnect." << endl;
+  }
+
+  setConnectionStatus(ConnectionStatus::NotConnected);
+
+  if(client_socket_) {
+    client_socket_->disconnectFromHost();
+  }
+}
+
+NodeTree* NetworkClient::nodeTree() {
+  return node_tree_;
+}
+
+unsigned int NetworkClient::protocolVersion() {
+  return protocol_version_;
+}
+
+QString NetworkClient::clientName() {
+  return client_name_;
+}
+
+QString NetworkClient::clientVersion() {
+  return client_version_;
+}
+
+QString NetworkClient::clientDescription() {
+  return client_description_;
+}
+
+QString NetworkClient::clientType() {
+  return client_type_;
+}
+
+QString NetworkClient::authenticationKey() {
+  return authentication_key_;
+}
+
+QTextStream* NetworkClient::output() {
+  return output_stream_;
+}
+
+void NetworkClient::setOutput(QTextStream* stream) {
+  output_stream_ = stream;
+}
+
+void NetworkClient::sendMsgConnect() {
+  std::shared_ptr<std::string> client_name_ptr(
+      new std::string(client_name_.toStdString()));
+  std::shared_ptr<std::string> client_version_ptr(
+      new std::string(client_version_.toStdString()));
+  std::shared_ptr<std::string> client_description_ptr(
+      new std::string(client_description_.toStdString()));
+  std::shared_ptr<std::string> client_type_ptr(
+      new std::string(client_type_.toStdString()));
+  std::shared_ptr<std::string> authentication_key_ptr(
+      new std::string);
+
+  msg_ptr msg(new messages::MsgConnect(
+      1,
+      pair_str(true, client_name_ptr),
+      pair_str(true, client_version_ptr),
+      pair_str(true, client_description_ptr),
+      pair_str(true, client_type_ptr),
+      pair_str(true, authentication_key_ptr)
+      ));
+
+  sendMessage(msg);
+}
+
+void NetworkClient::sendMessage(msg_ptr msg) {
+  if (client_socket_ && client_socket_->isValid()) {
+    msgpack::sbuffer buf;
+    msgpack::packer<msgpack::sbuffer> packer(buf);
+    messages::MsgpackWrapper::dumpObject(packer, msg);
+    client_socket_->write(buf.data(), buf.size());
+  }
+}
+
+void NetworkClient::registerMessageHandlers() {
+  message_handlers_["sub_cancelled"] = &NetworkClient::handleTreeModificationMessage;
```

Comment:
Why does this has the same handler as i.e. `get_reply` - this message doesn't correspond to any tree modification, and is only ack to request to sub cancellation.